File: src/test/java/shopping/domain/entity/ProductTest.java
package shopping.domain.entity;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import shopping.domain.vo.Image;
import shopping.domain.vo.Name;
import shopping.domain.vo.Price;

import static org.assertj.core.api.Assertions.assertThatNoException;

class ProductTest {
    @Test
    @DisplayName("상품은 id, 이름, 이미지, 가격 정보를 가지고 있다.")
    void createProduct() {
        assertThatNoException().isThrownBy(() -> new Product(1L, new Name("치킨"), new Image("chicken.png"), new Price(2000)));
    }
}

File: src/test/kotlin/shopping/ProductTest.kt
package shopping

import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.assertions.throwables.shouldThrow
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import shopping.domain.Product

class ProductTest {

    @Test
    fun `상품 이름은 공백을 포함하여 최대 15자까지이다`(){
        shouldNotThrowAny { ProductFixture.of("아이스아메리카노") }
        shouldThrow<IllegalArgumentException> { ProductFixture.of("아이스아메리카노아이스아메리카노카페라떼") }

    }

    @ValueSource(strings = ["[아메리카노]","에스프레소+카페라떼"])
    @ParameterizedTest
    fun `특수문자는 일부만 가능하다`(param: String) {
        shouldNotThrowAny { ProductFixture.of(param) }
    }

    @ValueSource(strings = ["*아메리카노*","<에스프레소+카페라떼>","★",","])
    @ParameterizedTest
    fun `특수문자는 일부 이외에는 불가능하다`(param: String) {
        shouldThrow<IllegalArgumentException> { ProductFixture.of(param) }
    }
}

class ProductFixture() {
    companion object {
        fun of(name: String): Product {
            return Product(1L, name, 0, "")
        }
    }
}
